tests/functional/conftest.py

```diff
@@ -1,100 +1,344 @@
-def add_proxies_to_products(api_products):
+def nhs_number(patient: Patient) -> str:
+    return patient.nhs_number
+
+
+@given("I have a patient's record to update", target_fixture='record_to_update')
+def record_to_update(update: Update, headers_with_authorization: dict, pds_url: str) -> dict:
+    response = retrieve_patient(headers_with_authorization, update.nhs_number, pds_url)
+
+    update.record_to_update = json.loads(response.text)
+    update.etag = response.headers['Etag']
+
+    return update.record_to_update
+
+
+@given("I wish to update the patient's gender")
+def add_new_gender_to_patch(update: Update) -> None:
+    current_gender = update.record_to_update['gender']
+    new_gender = 'male' if current_gender == 'female' else 'female'
+    update.value = new_gender
+
+
+@given(
+    parsers.cfparse(
+        "I don't have {_:String} {header_field:String} header",
+        extra_types=dict(String=str)
+    ),
+    target_fixture='headers_with_authorization'
+)
+def remove_header(headers_with_authorization, header_field) -> dict:
+    headers_with_authorization.pop(header_field)
+    return headers_with_authorization
+
+
+@given(
+    parsers.cfparse(
+        'I have a header {field:String} value of "{value:String}"',
+        extra_types=dict(String=str)
+    ),
+    target_fixture='headers_with_authorization')
+def update_header(headers_with_authorization: dict, field: str, value: str) -> dict:
+    headers_with_authorization.update({field: value})
+    return headers_with_authorization
+
+
+@given(
+    parsers.cfparse(
+        "I have an empty {field:String} header",
+        extra_types=dict(String=str)
+    ),
+    target_fixture='headers_with_authorization')
+def empty_header(headers_with_authorization: dict, field: str) -> dict:
+    headers_with_authorization.update({field: ''})
+    return headers_with_authorization
+
+
+@pytest.fixture
+def query_params(search: Search) -> str:
+    return urllib.parse.urlencode(search.query)
+
+
+@when('I retrieve a patient', target_fixture='response')
+def retrieve_patient(headers_with_authorization: dict, nhs_number: str, pds_url: str) -> Response:
+    return get(url=f"{pds_url}/Patient/{nhs_number}", headers=headers_with_authorization)
+
+
+@when('I retrieve their related person', target_fixture='response')
+def retrieve_related_person(headers_with_authorization: dict, nhs_number: str, pds_url: str) -> Response:
+    return get(url=f"{pds_url}/Patient/{nhs_number}/RelatedPerson", headers=headers_with_authorization)
+
+
+@when("I update the patient's PDS record", target_fixture='response')
+def update_patient(headers_with_authorization: dict, update: Update, pds_url: str) -> Response:
+    headers = headers_with_authorization
+    headers.update({
+        "Content-Type": "application/json-patch+json",
+        "If-Match": update.etag,
+    })
 
-    # Check if we need to add an extra proxy *-asid-required-* to the product used for testing
-    proxy_name = config.PROXY_NAME
```

Comment:
why are we removing logging?

Reply:
They were introduced as a debugging technique during the migration to Test Utils 2.0. As such, I thought it prudent to remove them. The logging only gets written to the pipeline ui, as such they make the pipeline output unnecessarily verbose and difficult to read.

Reply:
okay, that's reasonable